pkg/controller/agent/vlanconfig/controller.go

```diff
@@ -0,0 +1,402 @@
+		return false, "", nil
+	}
+
+	for _, n := range matchedNodes {
+		if n == h.nodeName {
+			node, err := h.nodeCache.Get(n)
+			if err != nil {
+				return false, "", err
+			}
+			key := network.GroupName + "/" + vc.Spec.ClusterNetwork
+			return true, node.Labels[key], nil
+		}
+	}
+
+	return false, "", nil
+}
+
+func (h Handler) setupVLAN(vc *networkv1.VlanConfig) error {
+	var v *vlan.Vlan
+	var setupErr error
+	var vids []uint16
+	var uplink *iface.Link
+	// get VIDs
+	vids, setupErr = h.getNadVidList(iface.GenerateName(vc.Spec.ClusterNetwork, iface.BridgeSuffix))
+	if setupErr != nil {
+		goto updateStatus
+	}
+	// construct uplink
+	uplink, setupErr = setUplink(vc)
+	if setupErr != nil {
+		goto updateStatus
+	}
+	// set up VLAN
+	v = vlan.NewVlan(vc.Spec.ClusterNetwork, vids)
+	setupErr = v.Setup(uplink)
+
+updateStatus:
+	// Update status and still return setup error if not nil
+	if err := h.updateStatus(vc, v, setupErr); err != nil {
+		return fmt.Errorf("update status into vlanstatus %s", h.statusName(vc.Name))
+	}
+	if setupErr != nil {
+		return fmt.Errorf("set up VLAN failed, vlanconfig: %s, node: %s, error: %w", vc.Name, h.nodeName, setupErr)
+	}
+	// update node labels for pod scheduling
+	if err := h.addNodeLabel(vc); err != nil {
+		return fmt.Errorf("add node label to node %s for vlanconfig %s failed, error: %w", h.nodeName, vc.Name, err)
+	}
+
+	return nil
+}
+
+func (h Handler) removeVLAN(vc *networkv1.VlanConfig) error {
+	var v *vlan.Vlan
+	var teardownErr error
+
+	v, teardownErr = vlan.GetVlan(vc.Spec.ClusterNetwork)
+	// We take it granted that `LinkNotFound` means the VLAN has been torn down.
+	if teardownErr != nil {
+		if errors.As(teardownErr, &netlink.LinkNotFoundError{}) {
+			teardownErr = nil
+		}
+		goto updateStatus
+	}
+	if teardownErr = v.Teardown(); teardownErr != nil {
+		goto updateStatus
+	}
+
+updateStatus:
+	if err := h.removeNodeLabel(vc); err != nil {
+		return err
+	}
+	if err := h.deleteStatus(vc, teardownErr); err != nil {
+		return fmt.Errorf("update status into vlanstatus %s", h.statusName(vc.Name))
+	}
+	if teardownErr != nil {
+		return fmt.Errorf("tear down VLAN failed, vlanconfig: %s, node: %s, error: %w", vc.Name, h.nodeName, teardownErr)
+	}
+
+	return nil
+}
+
+func setUplink(vc *networkv1.VlanConfig) (*iface.Link, error) {
+	// set link attributes
+	linkAttrs := netlink.NewLinkAttrs()
+	linkAttrs.Name = vc.Spec.ClusterNetwork + iface.BondSuffix
+	if vc.Spec.Uplink.LinkAttrs != nil {
+		if vc.Spec.Uplink.LinkAttrs.MTU != 0 {
+			linkAttrs.MTU = vc.Spec.Uplink.LinkAttrs.MTU
+		}
+		if vc.Spec.Uplink.LinkAttrs.TxQLen != 0 {
+			linkAttrs.TxQLen = vc.Spec.Uplink.LinkAttrs.TxQLen
+		}
+		if vc.Spec.Uplink.LinkAttrs.HardwareAddr != nil {
+			linkAttrs.HardwareAddr = vc.Spec.Uplink.LinkAttrs.HardwareAddr
+		}
+	}
+	// Note: do not use &netlink.Bond{}
+	bond := netlink.NewLinkBond(linkAttrs)
+	// set bonding mode
+	mode := netlink.BOND_MODE_BALANCE_TLB
```

Comment:
plz use active-backup as default
TLB has lots of problems with bridges

Reply:
@tjjh89017 Could you briefly add the issue TLB might cause here?

Reply:
TLB will modify the src mac for most packets for egress.
That will cause the switch will not detect some macs for a long time unless an ARP response is sent.
If MAC entry in the switch is expired, it will cause too much BUM traffic.